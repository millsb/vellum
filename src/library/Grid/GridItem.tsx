import * as React from "react";
import {reject, isNil, merge} from "ramda";
import {ReactChild, ReactElement, StatelessComponent} from "react";

export interface GridItemStyles {
    gridColumn?: string;
    gridRow?: string;
    gridArea?: string;
}

export interface GridItemProps {
    column?: string;
    row?: string;
    area?: string;
}

export const GridItem: StatelessComponent<GridItemProps> = ({children, row, column, area }) => {
    const gridStyles: GridItemStyles = reject(isNil, {
        gridArea: area,
        gridColumn: column,
        gridRow: row,
    });

    const renderChildren: any = () => {
        return React.Children.map(children, (child: ReactChild) => {
            if (typeof child !== "string") {
                const elChild = child as ReactElement<any>;
                const newStyles = merge(elChild.props.style, gridStyles);
                return React.cloneElement(elChild, { style: newStyles});
            } else {
                return child;
            }
        });
    };

    return (
        <React.Fragment>
            {renderChildren()}
        </React.Fragment>
    );
};

export default GridItem;
