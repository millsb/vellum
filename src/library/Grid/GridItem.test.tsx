import "jest";
import { configure, render } from "enzyme";
import * as React from "react";
import GridItem from "./GridItem";

// Configure enzyme with react 16 adapter
const Adapter: any = require("enzyme-adapter-react-16");
configure({ adapter: new Adapter() });

describe("GridItem", () => {
    it("should add inline styles with child", () => {
        const gridItem = render(<GridItem column={"1/3"} row={"2/2"}><div/></GridItem>);
        expect(gridItem).toMatchSnapshot();
    });

    it("should add merge existing styles with child", () => {
        const gridItem = render(<GridItem column={"1/3"} row={"2/2"}><div style={{color: "blue"}}/></GridItem>);
        expect(gridItem).toMatchSnapshot();
    });

    it("should assign fake subgrid styles when subgrid is selected", () => {
        const gridItem = render(<GridItem subgrid={true}><div/></GridItem>);
        expect(gridItem).toMatchSnapshot();
    });
});
