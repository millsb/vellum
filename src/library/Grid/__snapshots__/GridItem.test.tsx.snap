// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem should add inline styles with child 1`] = `
<div
  style="grid-columns:1/3;grid-rows:2/2"
/>
`;

exports[`GridItem should add merge existing styles with child 1`] = `
<div
  style="color:blue;grid-columns:1/3;grid-rows:2/2"
/>
`;

exports[`GridItem should assign fake subgrid styles when subgrid is selected 1`] = `
<div
  style="display:grid;grid-template-area:inherit;grid-template-column:inherit;grid-template-row:inherit"
/>
`;
